src/cedalion/sim/synthetic_artifact.py

```diff
@@ -0,0 +1,365 @@
+    """Returns list of channels involving selected optodes."""
+
+    sel_chan = []
+    for opt in optodes:
+        sel_chan.extend(ts.sel(channel=ts.source == opt).channel.values)
+        sel_chan.extend(ts.sel(channel=ts.detector == opt).channel.values)
+    sel_chan = np.unique(sel_chan).tolist()
+    return sel_chan
+
+
+def random_events_num(
+    time: xr.DataArray,
+    num_events: int,
+    types: list[str],
+    channels: list[str] | None = None,
+):
+    """Generates timing data for random events.
+
+    Events are randomly selected from the types list and assigned random onset/duration
+    within the time series.
+
+    Args:
+        time: Time series to which the events are added.
+        num_events: Number of events to generate.
+        types: List of event types.
+        channels: List of channels to which the events apply.
+
+    Returns:
+        DataFrame with event timing data (columns onset_time, duration, trial_type,
+        value, channel).
+    """
+
+    timing = pd.DataFrame(columns=TIMING_COLUMNS)
+
+    for i in range(num_events):
+        onset_time = np.random.uniform(time[0], time[-1])
+        duration = np.random.uniform(0.1, time[-1] - onset_time)
+        type = np.random.choice(types)
+        timing = add_event_timing([(onset_time, duration)], type, channels, timing)
+
+    return timing
+
+
+def random_events_perc(
+    time: xr.DataArray,
+    perc_events: float,
+    types: list[str],
+    channels: list[str] | None = None,
+    min_dur: float = 0.1,
+    max_dur: float = 0.4,
+    timing: pd.DataFrame | None = None,
+) -> pd.DataFrame:
+    """Generates timing data for random events. Mainly intended for spike artifacts.
+
+    Events are randomly selected from the types list and assigned random onset/duration
+    within the time series.
+
+    Args:
+        time: Time series to which the events are added.
+        perc_events: Percentage of time series to cover with events.
+        types: List of event types.
+        channels: List of channels to which the events apply.
+        min_dur: Minimum duration of the events.
+        max_dur: Maximum duration of the events.
+        timing (optional): DataFrame with existing event timing data.
+
+    Returns:
+        DataFrame with event timing data (columns onset_time, duration, trial_type,
+        value, channel).
+    """
+
+    if timing is None:
+        timing = pd.DataFrame(columns=TIMING_COLUMNS)
+
+    event_time = 0
+    total_time = time[-1].item() - time[0].item()
+
+    while event_time < total_time * perc_events:
+        onset_time = np.random.uniform(time[0], time[-1])
+        duration = np.random.uniform(min_dur, max_dur)
+        type = np.random.choice(types)
+        if onset_time + duration < time[-1]:
+            timing = add_event_timing([(onset_time, duration)], type, channels, timing)
+            event_time += duration
+    return timing
+
+
+def sliding_window(ts: cdt.NDTimeSeries, window_size, step_size):
```

Comment:
Could this be simplified by using DataArray.rolling?

Reply:
This has already been done (in add_artifacts), I just forgot to remove the old functions